pkg/kubelet/allocation/allocation_manager.go

```diff
@@ -114,6 +133,91 @@ func (m *manager) UpdatePodFromAllocation(pod *v1.Pod) (*v1.Pod, bool) {
 	return updatePodFromAllocation(pod, allocs)
 }
 
+func GetPodResizeAllocation(pod *v1.Pod) (*PodResourceSummary, bool) {
+	resizeManager.mutex.RLock()
+	defer resizeManager.mutex.RUnlock()
+
+	resizeAllocation, exists := resizeManager.podResizeAllocation[pod.UID]
+	if !exists {
+		return nil, false
+	}
+	return resizeAllocation, true
+}
+
+func ClearPodResizeAllocation(pod *v1.Pod) {
+	resizeManager.mutex.Lock()
+	defer resizeManager.mutex.Unlock()
+	delete(resizeManager.podResizeAllocation, pod.UID)
+}
+
+// Compare reqOld and reqNew, if different, store reqNew resource to diffReq
+func compareAndSaveDiff(reqOld, reqNew v1.ResourceRequirements) v1.ResourceRequirements {
+	var diffReq v1.ResourceRequirements
+
+	// Compare Requests resource
+	diffReq.Requests = make(v1.ResourceList)
+	for key, reqNewVal := range reqNew.Requests {
+		if reqOldVal, exists := reqOld.Requests[key]; !exists || !reqOldVal.Equal(reqNewVal) {
+			diffReq.Requests[key] = reqNewVal
+		}
+	}
+
+	// Compare Limits resource
+	diffReq.Limits = make(v1.ResourceList)
+	for key, reqNewVal := range reqNew.Limits {
+		if reqOldVal, exists := reqOld.Limits[key]; !exists || !reqOldVal.Equal(reqNewVal) {
+			diffReq.Limits[key] = reqNewVal
+		}
+	}
+	return diffReq
+}
+
+func updateContainerResizeAllocation(pod *v1.Pod, c v1.Container, diffReq v1.ResourceRequirements) *PodResourceSummary {
+	podResizeAlloc, exists := GetPodResizeAllocation(pod)
+	if !exists {
+		podResizeAlloc = &PodResourceSummary{}
+	}
+
+	// Check initcontainer
+	isInitContainer := false
+	for _, initContainer := range pod.Spec.InitContainers {
+		if c.Name == initContainer.Name {
+			isInitContainer = true
+			break
+		}
+	}
+
+	if isInitContainer {
+		if utilfeature.DefaultFeatureGate.Enabled(features.SidecarContainers) && podutil.IsRestartableInitContainer(&c) {
+			if podResizeAlloc.InitContainers == nil {
+				podResizeAlloc.InitContainers = make(map[string]v1.ResourceRequirements)
+			}
+			podResizeAlloc.InitContainers[c.Name] = diffReq
+		}
+	} else {
```

Comment:
There are also efemeral containers, so if some container is not an init container, doesn't mean it's a regular container.